tsdb/exemplar.go

```diff
@@ -0,0 +1,216 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tsdb
+
+import (
+	"context"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/pkg/exemplar"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage"
+)
+
+type exemplarMetrics struct {
+	outOfOrderExemplars prometheus.Counter
+}
+
+func newExemplarMetrics(r prometheus.Registerer) *exemplarMetrics {
+	m := &exemplarMetrics{
+		outOfOrderExemplars: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "prometheus_tsdb_exemplar_out_of_order_exemplars_total",
+			Help: "Total number of out of order exemplar ingestion failed attempts",
+		}),
+	}
+	if r != nil {
+		r.MustRegister(m.outOfOrderExemplars)
+	}
+	return m
+}
+
+type CircularExemplarStorage struct {
+	metrics   *exemplarMetrics
+	lock      sync.RWMutex
+	exemplars []*circularBufferEntry
+	nextIndex int
+
+	// Map of series labels as a string to index entry, which points to the first
+	// and last exemplar for the series the exemplars circular buffer.
+	index map[string]*indexEntry
+}
+
+type indexEntry struct {
+	first int
+	last  int
+}
+
+type circularBufferEntry struct {
+	exemplar     exemplar.Exemplar
+	seriesLabels labels.Labels
+	next         int
+}
+
+// If we assume the average case 95 bytes per exemplar we can fit 5651272 exemplars in
+// 1GB of extra memory, accounting for the fact that this is heap allocated space.
+// If len < 1, then the exemplar storage is disabled.
+func NewCircularExemplarStorage(len int, reg prometheus.Registerer) (storage.ExemplarStorage, error) {
+	if len < 1 {
+		return &noopExemplarStorage{}, nil
+	}
+	return &CircularExemplarStorage{
+		exemplars: make([]*circularBufferEntry, len),
+		index:     make(map[string]*indexEntry),
+		metrics:   newExemplarMetrics(reg),
+	}, nil
+}
+
+func (ce *CircularExemplarStorage) Appender() storage.ExemplarAppender {
+	return ce
+}
+
+func (ce *CircularExemplarStorage) ExemplarAppender() storage.ExemplarAppender {
+	return ce
+}
+
+func (ce *CircularExemplarStorage) ExemplarQuerier(_ context.Context) (storage.ExemplarQuerier, error) {
+	return ce, nil
+}
+
+func (ce *CircularExemplarStorage) Querier(ctx context.Context) (storage.ExemplarQuerier, error) {
+	return ce, nil
+}
+
+// Select returns exemplars for a given set of series labels hash.
+func (ce *CircularExemplarStorage) Select(start, end int64, matchers ...[]*labels.Matcher) ([]exemplar.QueryResult, error) {
+	var ret []exemplar.QueryResult
+
+	ce.lock.RLock()
+	defer ce.lock.RUnlock()
+
+	// Checking against all exemplars.
+	for _, idx := range ce.index {
+		var se exemplar.QueryResult
+		e := ce.exemplars[idx.first]
+		se.SeriesLabels = e.seriesLabels
+		for e.exemplar.Ts <= end {
```

Comment:
I don't think we're guaranteed that the circular buffer containers exemplars in order...

Reply:
For a given series, it will be in order.

Reply:
Ah yes, fair enough - this code is pretty tricky to follow.  A comment would be helpful.

Reply:
I've added some comments.